### ISSUE TITLE
[doc] perlxstut update

### ISSUE THREAD

**Issue by ghost** [Perl/perl5#18163]
https://perldoc.perl.org/perlxstut.html

Perl 5.22.2

Running `h2xs -A -n Mytest` creates a `t/Mytest.t` file with `use strict;`.

`EXAMPLE 3`
    
```
$i = -1.5; &Mytest::round($i); is( $i, -2.0 );
$i = -1.1; &Mytest::round($i); is( $i, -1.0 );
$i = 0.0; &Mytest::round($i);  is( $i,  0.0 );
$i = 0.5; &Mytest::round($i);  is( $i,  1.0 );
$i = 1.2; &Mytest::round($i);  is( $i,  1.0 );
```

Running `make test` then of course outputs the question `(did you forget to declare "my $i"?)`.

The maintainer's e-mail address is furthermore invalid.

**Comment by ghost** [comment]
I'll check the other examples.

**Comment by ghost** [comment]
EXAMPLE 4

The content #include "./mylib.h" can be changed to #include "mylib.h".

The content ./Mytest2/mylib/mylib.h can be changed to Mytest2/mylib/mylib.h.

'NAME', 'VERSION_FROM', 'LIBS', 'DEFINE', 'INC' and 'MYEXTLIB' do not need to be enclosed in single quotes.

There's no single line that says "mylib" in the generated MANIFEST file.

```
bash-4.3$ cat Mytest2/MANIFEST
Changes
Makefile.PL
MANIFEST
Mytest2.xs
ppport.h
README
t/Mytest2.t
fallback/const-c.inc
fallback/const-xs.inc
lib/Mytest2.pm
bash-4.3$ 
```

The recommendation to edit the .pm file is obscure.

```
bash-4.3$ cat Mytest2/lib/Mytest2.pm | perl -ne 'print if $. == 13..28'
# Items to export into callers namespace by default. Note: do not export
# names by default without a very good reason. Use EXPORT_OK instead.
# Do not simply export all your public functions/methods/constants.

# This allows declaration       use Mytest2 ':all';
# If you do not need this, moving things directly into @EXPORT or @EXPORT_OK
# will save memory.
our %EXPORT_TAGS = ( 'all' => [ qw(
        TESTVAL
) ] );

our @EXPORT_OK = ( @{ $EXPORT_TAGS{'all'} } );

our @EXPORT = qw(
        TESTVAL
);
bash-4.3$ 
```

The number of tests in the Mytest2.t script should be changed to "5".

```
bash-4.3$ grep tests Mytest2/t/Mytest2.t
# change 'tests => 2' to 'tests => last_test_to_print';
use Test::More tests => 2;
bash-4.3$ 
```

**Comment by ghost** [comment]
EXAMPLE 5

`(did you forget to declare "my @a"?)`.

Include `my @a;` followed by a blank line at the top of the code segment to be added to Mytest.t.

<hr>

EXAMPLE 6

`(did you forget to declare "my $results"?)`

Include `my $results;` followed by a blank line at the top of the code segment to be added to Mytest.t.

<hr>

EXAMPLE 9 Passing open files to XSes

Is perlioputs a misprint?

These comments might be added, respectively from [perlapio](https://perldoc.perl.org/perlapio.html) and [perlguts](https://perldoc.perl.org/perlguts.html).

```
#define PERLIO_NOT_STDIO 0      /* For co-existence with stdio only */
#define PERL_NO_GET_CONTEXT     /* we want efficiency */
```

Suggested addition below this line.

The real work is done in the standard typemap.

For more details, see [perlapio](https://perldoc.perl.org/perlapio.html#Co-existence-with-stdio).

<hr>

See also

Perhaps add [perlapio](https://perldoc.perl.org/perlapio.html) to the list of URLs to consult.

**Comment by xsawyerx** [comment]
Thank you for checking the rest of the examples.

I made a few changes.

> EXAMPLE 4
> 
> The content #include "./mylib.h" can be changed to #include "mylib.h".
> 
> The content ./Mytest2/mylib/mylib.h can be changed to Mytest2/mylib/mylib.h.

This might have been the author trying to indicate more strongly that these files are in the current directory, but I revised this nonetheless.

> 'NAME', 'VERSION_FROM', 'LIBS', 'DEFINE', 'INC' and 'MYEXTLIB' do not need to be enclosed in single quotes.

They don't need to be, but I prefer them there nowadays. Autoquoting rules in Perl are... subtle (some sometimes strange), so it doesn't hurt to be more explicit as a practice.

> There's no single line that says "mylib" in the generated MANIFEST file.

I think this should be edited the code, rather than the tutorial.

> The recommendation to edit the .pm file is obscure.

Same as above.

> The number of tests in the Mytest2.t script should be changed to "5".

Same as above.

**Comment by xsawyerx** [comment]
> EXAMPLE 5
> 
> `(did you forget to declare "my @a"?)`.

Fixed!

> EXAMPLE 6
> 
> `(did you forget to declare "my $results"?)`

Fixed!

> EXAMPLE 9 Passing open files to XSes
> 
> Is perlioputs a misprint?

I think the idea here was to give an example of what would it look like if you implemented a function `perlioputs` to access `PerlIO_puts` from XS.

> These comments might be added, respectively from [perlapio](https://perldoc.perl.org/perlapio.html) and [perlguts](https://perldoc.perl.org/perlguts.html).
> 
> ```
> #define PERLIO_NOT_STDIO 0      /* For co-existence with stdio only */
> #define PERL_NO_GET_CONTEXT     /* we want efficiency */
> ```

While I think they require reading more to fully understand these comments, I'm in favor of putting them here so they don't make the user immediately wonder "oh wait... what are these?" This small hint tells them in a few words what these are for and they can decide for themselves when to read more about them.

> Suggested addition below this line.
> 
> The real work is done in the standard typemap.
> 
> For more details, see [perlapio](https://perldoc.perl.org/perlapio.html#Co-existence-with-stdio).

Done.

> See also
> 
> Perhaps add [perlapio](https://perldoc.perl.org/perlapio.html) to the list of URLs to consult.

Done.

**Comment by xsawyerx** [comment]
And I also removed calling functions with `&`. Thought about it more and dislike it enough to remove that part.

**Comment by ghost** [comment]
EXAMPLE 4

> They don't need to be, but I prefer them there nowadays.

I favour consistency with the Makefile.PL presentation from EXAMPLES 1 and 4, on the former NAME, VERSION_FROM, LIBS, DEFINE, and INC, are not enclosed with single quotes.

**Comment by Grinnz** [comment]
> > 'NAME', 'VERSION_FROM', 'LIBS', 'DEFINE', 'INC' and 'MYEXTLIB' do not need to be enclosed in single quotes.
> 
> They don't need to be, but I prefer them there nowadays. Autoquoting rules in Perl are... subtle (some sometimes strange), so it doesn't hurt to be more explicit as a practice.

They are actually extremely simple (though I empathize as I shared this view some years ago) - I think it's a failing of the documentation that it doesn't make this more obvious, but I can't say exactly what or where it should be improved to address this.

The best explanation of it currently is https://perldoc.pl/perldata#Identifier-parsing as it is the same as the identifier parsing rules (minus double colons), and boils down to "containing only letters, digits, or underscore, and not starting with a digit".

The leading minus is a weird addition to this since it is actually the unary minus operator applying its own same autoquoting rule to the following bareword.

**Comment by xsawyerx** [comment]
> > > 'NAME', 'VERSION_FROM', 'LIBS', 'DEFINE', 'INC' and 'MYEXTLIB' do not need to be enclosed in single quotes.
> > 
> > 
> > They don't need to be, but I prefer them there nowadays. Autoquoting rules in Perl are... subtle (some sometimes strange), so it doesn't hurt to be more explicit as a practice.
> 
> They are actually extremely simple (though I empathize as I shared this view some years ago) - I think it's a failing of the documentation that it doesn't make this more obvious, but I can't say exactly what or where it should be improved to address this.
> 
> The best explanation of it currently is https://perldoc.pl/perldata#Identifier-parsing as it is the same as the identifier parsing rules (minus double colons), and boils down to "containing only letters, digits, or underscore, and not starting with a digit".
> 
> The leading minus is a weird addition to this since it is actually the unary minus operator applying its own same autoquoting rule to the following bareword.

This misses a subtle point of newlines breaking autoquoting.

**Comment by xsawyerx** [comment]
> EXAMPLE 4
> 
> > They don't need to be, but I prefer them there nowadays.
> 
> I favour consistency with the Makefile.PL presentation from EXAMPLES 1 and 4, on the former NAME, VERSION_FROM, LIBS, DEFINE, and INC, are not enclosed with single quotes.

I just pushed a fix to the PR that normalizes the quoting for other examples of `WriteMakefile` function calls.

**Comment by ghost** [comment]
@Grinnz would you concur of the commits that `h2xs` generated files' embellishment with single quotes and comments is justifiable for the purposes of clarity?

**Comment by Grinnz** [comment]
I don't think either way is any more clear, but the quotes add clutter, I am not sure how relevant my personal style preferences are to this document though.

**Comment by ghost** [comment]
There's something to be said about leaving program, in this instance `h2xs`, output as rendered.

EXAMPLE 9 Passing open files to XSes

The basis for my suggested changes was due to PERL_NO_GET_CONTEXT already being addressed in SPECIAL NOTES.

Factually PERL_NO_GET_CONTEXT is in the `h2xs` generated *.xs files whereas PERLIO_NOT_STDIO 0 was added in the referenced example.

**Comment by xsawyerx** [comment]
> There's something to be said about leaving program, in this instance `h2xs`, output as rendered.

True.


> EXAMPLE 9 Passing open files to XSes
> 
> The basis for my suggested changes was due to PERL_NO_GET_CONTEXT already being addressed in SPECIAL NOTES.
> 
> Factually PERL_NO_GET_CONTEXT is in the `h2xs` generated *.xs files whereas PERLIO_NOT_STDIO 0 was added in the referenced example.

Are you asking for a particular change here? I don't follow.

**Comment by ghost** [comment]
> Are you asking for a particular change here?

SHORT

In this instance the reader may wonder why the single quotes are in the tutorial but absent in the files generated by the command the reader is told to run.

I'm not fond of the proposed addition of the single quotes.

<hr>

TL;DR

My preferences are not to oblige the reader to think more.

EXAMPLE 4

Telling the reader to replace a non-existant line with three lines rather than to append three lines.

Telling the reader to edit the .pm file and change the variable `@EXPORT` to `@EXPORT_OK`.

Both are already present in generated *.pm files.

Running `h2xs -A -n Mytest` versus `h2xs -O -n Mytest2 ./Mytest2/mylib/mylib.h` generates different *.pm content; below after running the `h2xs -A -n Mytest` `@EXPORT` is empty, this is not the case in the Mytest2.pm file.

```
bash-4.3$ cat Mytest/lib/Mytest.pm | perl -ne 'print if $. == 11..26'
# Items to export into callers namespace by default. Note: do not export
# names by default without a very good reason. Use EXPORT_OK instead.
# Do not simply export all your public functions/methods/constants.

# This allows declaration	use Mytest ':all';
# If you do not need this, moving things directly into @EXPORT or @EXPORT_OK
# will save memory.
our %EXPORT_TAGS = ( 'all' => [ qw(
	
) ] );

our @EXPORT_OK = ( @{ $EXPORT_TAGS{'all'} } );

our @EXPORT = qw(
	
);
bash-4.3$ 
```

Telling the reader to change the number of tests to "4" when the test fails unless changed to "5".

**Comment by apparluk** [comment]
Running make creates mylib/libmylib.a, so there's no need to "cd into the mylib directory and run make in there as well". 

EXAMPLE 4

"Now run perl on the top-level Makefile.PL. Notice that it also created a Makefile in the mylib directory. Run make and watch that it does cd into the mylib directory and run make in there as well."

**Comment by apparluk** [comment]
Did the initial h2xs implementation generate  *.xs content as #define PERL_NO_GET_CONTEXT?

An update of John Keiser's (2001 circa) code.

https://github.com/nurmi-labs/blogger/wiki/perlxs

**Comment by apparluk** [comment]
Peripheral to this ticket but perhaps worth mention regarding [perlxs](https://perldoc.perl.org/5.32.0/perlxs.html) examples.

Makefile.PL

```
LIBS              => ['-ltirpc'], # e.g., '-lm'
DEFINE            => '', # e.g., '-DHAVE_SOMETHING'
INC               => '-I. -I/usr/include/tirpc', # e.g., '-I. -I/usr/include/other'

```
RPC.xs

```
- #include <rpc/rpc.h>
+ #include <tirpc/rpc/rpc.h>

```
```
bash-4.3$ grep -nr rpcb_gettime /usr/include/tirpc
/usr/include/tirpc/rpc/rpcb_clnt.h:49: *	success = rpcb_gettime(host, timep)
/usr/include/tirpc/rpc/rpcb_clnt.h:80:extern bool_t rpcb_gettime(const char *, time_t *);
bash-4.3$ grep -nr getnetconfigent /usr/include/tirpc
/usr/include/tirpc/netconfig.h:85:extern struct netconfig *getnetconfigent (const char *);
bash-4.3$ 
```

**Comment by apparluk** [comment]
For pædogogical reasons embellish code.

Otherwise if code runs do not mess with it.

**Comment by apparluk** [comment]
Not implementing https://github.com/Perl/perl5/pull/18165/commits/e3e54b068bf3216644e75efd29d8c181e6868428 and limiting the proposed changes in https://github.com/Perl/perl5/commit/088fd70e7f982a31bf590679ffde65cac73c5872 to `my $i;` followed by a blank line, would be a perhaps not so obvious clue to the reader as to either the original author's writing style or the time period of its publication.

I would suggest under Author deleting the e-mail address and adding a line (as below) and under Last Changed revising YYYY-MM-DD.

```
Author

Jeff Okamoto <okamoto@corp.hp.com>

Reviewed and assisted by Dean Roehrich, Ilya Zakharevich, Andreas Koenig, and Tim Bunce.

PerlIO material contributed by Lupe Christoph, with some clarification by Nick Ing-Simmons.

Changes for h2xs as of Perl 5.8.x by Renee Baecker

It is now maintained as part of Perl itself.

Last Changed

2012-01-20
```

**Comment by ghost** [comment]
<s>`https://github.com/vaerksted/perl5`</s> deleted Oct 22, 2020

**Comment by jkeenan** [comment]
> https://github.com/vaerksted/perl5
> patch for perlxstut

This is not the best way to submit a request for changes to the Perl 5 core distribution.

While there are a number of ways to create and submit a patch, since you have already opened a GH issue I'll only discuss the way that involve github.com, namely, a pull request.

You should fork the entire repository to your github account, clone to your local machine, create a branch to hold your revisions on your local machine, edit, make sure the entire test suite passes, push your branch to your github repository, and then initiate a pull request.  The Perl-specific details of this can be found by calling `perldoc perlhack`; the Help section of github.com should also be consulted.

What you appear to have done above is to create a new repository in your github account that contains only the particular file you want to revise and a patch of that file.  While we do accept single patches provided as email attachments (and we have since 1987!), if you use github you should follow the standard pull request procedures.

Thank you very much.
Jim Keenan

**Comment by xsawyerx** [comment]
I've just applied all of the remaining suggestions and will merge them, closing this ticket.

If you have any other suggestions, I highly recommend submitting them as either a patch file or a GitHub Pull Request. If you need assistance in creating your first Pull Request, I recommend reading [this](https://opensource.com/article/19/7/create-pull-request-github) or [this](https://docs.github.com/en/free-pro-team@latest/github/collaborating-with-issues-and-pull-requests/creating-a-pull-request). The above guide by @jkeenan is also very helpful.

Thank you for all the help. :)

**Comment by apparluk** [comment]
After declaring `"my @a"` I neglected to add a blank line. 

```
	my @a;
	@a = &Mytest::statfs("/blech");
	ok( scalar(@a) == 1 && $a[0] == 2 );
	@a = &Mytest::statfs("/");
	is( scalar(@a), 7 );
```

**Comment by apparluk** [comment]
@jkeenan

Suggested change in perlhack.pod.

replace

`  % git remote add fork git@github.com:MyUser/perl5.git`

with

`  % git remote add fork https://github.com/MyUser/perl5.git`

<hr>

_cf_: https://github.com/Perl/perl5/issues/18163#issuecomment-699858249

@xsawyerx 

Where are rpcb_gettime and getnetconfigent defined on your hardware?

**Comment by apparluk** [comment]
Citing v5.22.2 this ticket was was opened regarding failing tests.

It was neither my intention to modernise code segment style nor textual content.

Perl 4 is appears in SPECIAL NOTES.

Some GNU Linux distributions do not include libperl.a.

Code segment modernisation should be contemporaneous with textual modernisation (in the same ticket/PR).

<hr>

@xsawyerx

Noting you're affiliation with Perl 7 development I think the religious connotations should be purged from the language.

**Comment by apparluk** [comment]
[perl #121223] explain PERL_NO_GET_CONTEXT a little 
committed Feb 16, 2014

add #define PERL_NO_GET_CONTEXT to ParseXS's pods
committed Feb 16, 2014 

It seemed justifiable regarding PERLIO_NOT_STDIO 0 to add the 2 perlapio links in EXAMPLE 9.

![v5 6 1](https://user-images.githubusercontent.com/52227507/95824238-43613480-0d2f-11eb-915a-5161ec7419e0.jpg)

_cf_: https://github.com/Perl/perl5/pull/18205#discussion_r499306286

@Grinnz 

In the circa v5.6.1 output `-I.` doesn't appear  in the `h2xs -A -n Mytest` output Makefile.PL.
